### PR TITLE
[FLINK-3633] Fix user code de/serialization in ExecutionConfig

### PR DESCRIPTION
**Pull request by tillrohrmann**
FLINK-3327 moved the ExecutionConfig directly to the JobGraph so that it was serialized
and deserialized using the system class loader when sending a SubmitJob message to the
JobManager. This is problematic since the ExecutionConfig can contain user code class
which require a user code class loader for deserialization. In order to circumvent the
problem, a UserCodeValue class was introduced which automatically sends the wrapped value
as a byte array. On the receiving side, the wrapped value has to be explicitly deserialized
providing a class loader.

To test the feature the ScalaShellITCase.testSubmissionOfExternalLibrary was adapted
to register org.apache.flink.ml.math.Vector at the ExecutionConfig.

This commit also re-introduces the removed ExecutionConfig.CONFIG_KEY key, so that
version 1.1 does not break the API.

### CODE REVIEW

File: docs/setup/config.md [hunk 1]
```diff
@@ -183,7 +184,7 @@ The following parameters configure Flink's JobManager and TaskManagers.
 - `akka.transport.threshold`: Threshold for the transport failure detector. Since Flink uses TCP, the detector is not necessary and, thus, the threshold is set to a high value (DEFAULT: **300**).
 - `akka.tcp.timeout`: Timeout for all outbound connections. If you should experience problems with connecting to a TaskManager due to a slow network, you should increase this value (DEFAULT: **akka.ask.timeout**).
 - `akka.throughput`: Number of messages that are processed in a batch before returning the thread to the pool. Low values denote a fair scheduling whereas high values can increase the performance at the cost of unfairness (DEFAULT: **15**).
-- `akka.log.lifecycle.events`: Turns on the Akka's remote logging of events. Set this value to 'on' in case of debugging (DEFAULT: **off**).
+- `akka.log.lifecycle.events`: Turns on the Akka's remote logging of events. Set this value to 'on' in case of debugging (DEFAULT: **false**).
```
Comment (greghogan) [inline docs/setup/config.md]: Should this say, "Set this value to 'true' in case ..."?

Comment (tillrohrmann) [inline docs/setup/config.md]: Yes, definitely. Thanks for spotting it :-)

File: flink-core/src/main/java/org/apache/flink/util/UserCodeValue.java [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+	private byte[] serializedValue;
+
+	public UserCodeValue(T value) {
+		this.value = value;
+	}
+
+	public T getValue() {
+		if (value == null && serializedValue != null) {
+			throw new NullPointerException("The user code value has not been serialied. Call " +
```
Comment (greghogan) [inline flink-core/src/main/java/org/apache/flink/util/UserCodeValue.java]: "serialized" -> "deserialized"?

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java [hunk 3]
```diff
@@ -499,7 +514,7 @@ else if (current == ExecutionState.CANCELING) {
 					jobId, vertexId, executionId, userCodeClassLoader, actorAskTimeout);
 
 			Environment env = new RuntimeEnvironment(jobId, vertexId, executionId,
-					executionConfig, taskInfo, jobConfiguration, taskConfiguration,
+				executionConfig, taskInfo, jobConfiguration, taskConfiguration,
```
Comment (greghogan) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: Formatting regression?

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java [hunk 4]
```diff
@@ -109,22 +113,22 @@
...
+	private final UserCodeValue<LinkedHashMap<Class<?>, SerializableSerializer<?>>> defaultKryoSerializers = new UserCodeValue<>(new LinkedHashMap<Class<?>, SerializableSerializer<?>>());
 
-	private final LinkedHashMap<Class<?>, Class<? extends Serializer<?>>> defaultKryoSerializerClasses = new LinkedHashMap<>();
+	private final UserCodeValue<LinkedHashMap<Class<?>, Class<? extends Serializer<?>>>> defaultKryoSerializerClasses = new UserCodeValue<>(new LinkedHashMap<Class<?>, Class<? extends Serializer<?>>>());
 
-	private final LinkedHashSet<Class<?>> registeredKryoTypes = new LinkedHashSet<>();
+	private final UserCodeValue<LinkedHashSet<Class<?>>> registeredKryoTypes = new UserCodeValue<>(new LinkedHashSet<Class<?>>());
 
-	private final LinkedHashSet<Class<?>> registeredPojoTypes = new LinkedHashSet<>();
+	private final UserCodeValue<LinkedHashSet<Class<?>>> registeredPojoTypes = new UserCodeValue<>(new LinkedHashSet<Class<?>>());
```
Comment (rmetzger) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java]: You forgot to put the `GlobalJobParameters` into a `UserCodeValue`. That one can also contain usercode.

File: flink-core/src/main/java/org/apache/flink/util/UserCodeValue.java (hunk 2 above)
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/util/UserCodeValue.java]: good catch. Thanks :-)

File: flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java (hunk 3 above)
Comment (tillrohrmann) [inline flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java]: True, will revert it.

File: flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java (hunk 4 above)
Comment (tillrohrmann) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java]: That is true, I will fix it. Personally, I think this `GlobalJobParameters` is a huge crime. It basically allows you to store anything in the `ExecutionConfig`. Sometimes I'm overcome with the feeling that this `ExecutionConfig` is simply a dumping ground for all kinds of parameters which are somehow needed somewhere.

Comment (rmetzger) [inline flink-core/src/main/java/org/apache/flink/api/common/ExecutionConfig.java]: Thanks for the fix.

The original motivation for adding it was the following:
Users requested a way to have one central configuration file for all their job's UDFs. With setting global job parameters (for example the ParameterTool), they can use a properties file and the KVs from there are accessible from all UDFs.
Also, the same user wanted to have these configuration properties accessible in the webinterface. So if the `toMap()` method is overwritten (the ParameterTool is doing that), the job parameters are accessible from the web interface.

**Comment by tillrohrmann** [comment]
Due to the tight coupling of the ExecutionConfig and multiple Flink components
(e.g. PojoSerializer) the automatic serialization and manual deserialization of
user code objects via the UserCodeValue class caused problems. In order to minimize
the impact of the changes, I changed the serialization strategy to an explicit one.
One has to call `ExecutionConfig.serializeUserCode` to store the user code objects in a SerializedValue
object and nulling the corresponding member fields. If that is not done, then it is
assumed that the object is deserialized using a user code class loader. On the receiving side one has to call `ExecutionConfig.deserializeConfig` providing a class loader which knows the user code classes.

**Comment by rmetzger** [comment]
The issue has been introduced after the 1.0 release, so we don't need to incldue this into 1.0.1?

**Comment by tillrohrmann** [comment]
Yes I think so.

**Comment by rmetzger** [comment]
okay, thank you.

**Comment by uce** [comment]
The changes look good and make parts where user code is serialized very clear. :+1: 

I've verified that the `ScalaShellITCase` with the change in this PR fails for the current master. I'm wondering whether it is possible to make the test for this more explicit by adding a JAR with the example program outlined in the JIRA issue to the `ClassLoaderITCase`.

In any case, +1 to merge. It's your call whether you add a further class loader test or not.

**Comment by tillrohrmann** [comment]
I agree that having a dedicated class loading test would be nicer. However, we would need to build a jar containing the test job before executing the test, right? So the easiest solution for that would be a new module. But that is also really ugly. @uce do you have a better idea for that?

**Comment by uce** [comment]
The other programs that are executed in `ClassLoaderITCase` are contained in the `org.apache.flink.test.classloading.jar` package (search for `KMeansForTest` for example). They are then packaged by the assembly plugin (see `pom.xml` of `flink-tests`). The `ClassLoaderITCase` sets up a cluster and submits the assembled JARs.

Can we just follow this setup?

**Comment by tillrohrmann** [comment]
That is a good idea @uce :-) I'll add a corresponding test for the `ExecutionConfig`.

**Comment by tillrohrmann** [comment]
I've added an additional test case for registered user code types in the `ExecutionConfig`. I will rebase this PR and if travis gives green light, then I'll merge it.

**Comment by tillrohrmann** [comment]
Failing test case seems to be unrelated

**Pull request closed without merging**